flow/layers/layer.h

```diff
@@ -145,14 +145,46 @@ class Layer {
 
   bool needs_painting() const { return !paint_bounds_.isEmpty(); }
 
+  // True iff the layer, or some descendant of the layer, performs an
+  // operation which depends on (i.e. must read back from) the surface
+  // on which it is rendered in any way beyond the functionality of
+  // BlendMode. This value has no setter as it is computed from other
+  // flags and properties on the layer.
+  //
+  // see Layer::UpdateTreeReadsSurface()
+  // see Layer::set_layer_reads_surface()
+  // see ContainerLayer::set_renders_to_save_layer()
+  // see ContainerLayer::UpdateChildReadback()
+  bool tree_reads_surface() const { return tree_reads_surface_; }
+
   uint64_t unique_id() const { return unique_id_; }
 
+ protected:
+  // Compute a new value for tree_reads_surface_ from all of the various
+  // properties of this layer.
+  // Used by UpdateTreeReadsSurface()
+  virtual bool ComputeTreeReadsSurface() const;
```

Comment:
I'm Ok with the approach here. The following are some thoughts for future cleanups:

It seems to me that `ComputeTreeReadsSurface` or `TreeReadsSurface` is only needed by `ContainerLayer`. There's no leaf `Layer` that reads surface. The current approach of defining a virtual function is probably needed because we made `ContainerLayer` and `Layer` two classes. I feel that we can save some code (in this and probably some other cases) if we merge `ContainerLayer` and `Layer` into a single class (e.g., a leaf `Layer` is simply a `ContainerLayer` with no children; the memory overhead of having an empty `std::vector` should be minimal). @chinmaygarde : what's your thought on this?

Reply:
I'd rather spend time on unraveling the surface decision so all of this could move into PreRoll, then all of these micro-improvements would become irrelevant.

Note that I actually considered just having all of the flags in the base Layer class anyway, but decided to follow the distinction of the class hierarchy, so that almost happened anyway...